File: MVVMJavaRx/app/src/debug/java/com/vascome/fogtail/presentation/dev_settings/presenters/BasePresenter.java
package com.vascome.fogtail.presentation.dev_settings.presenters;

import android.support.annotation.CallSuper;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;

/**
 * Created by vasilypopov on 12/17/17
 * Copyright (c) 2017 MVVMJavaRx. All rights reserved.
 */

public abstract class BasePresenter<V> {

    @Nullable
    private volatile V view;

    @CallSuper
    public void bindView(@NonNull V view) {
        final V previousView = this.view;

        if (previousView != null) {
            throw new IllegalStateException("Previous view is not unbounded! previousView = " + previousView);
        }

        this.view = view;
    }

    @Nullable
    protected V view() {
        return view;
    }

    @CallSuper
    @SuppressWarnings("PMD.CompareObjectsWithEquals")
    public void unbindView(@NonNull V view) {
        final V previousView = this.view;

        if (previousView == view) {
            this.view = null;
        } else {
            throw new IllegalStateException("Unexpected view! previousView = " + previousView + ", view to unbind = " + view);
        }
    }
}

File: MVVMKotlin/app/src/debug/java/com/vascome/fogtail/presentation/dev_settings/presenters/BasePresenter.kt
package com.vascome.fogtail.presentation.dev_settings.presenters

import android.support.annotation.CallSuper

/**
 * Created by vasilypopov on 12/17/17
 * Copyright (c) 2017 MVVMJavaRx. All rights reserved.
 */

abstract class BasePresenter<V> {

    @Volatile private var view: V? = null

    @CallSuper
    open fun bindView(view: V) {
        val previousView = this.view

        if (previousView != null) {
            throw IllegalStateException("Previous view is not unbounded! previousView = " + previousView)
        }

        this.view = view
    }

    fun view(): V? {
        return view
    }

    @CallSuper
    fun unbindView(view: V) {
        val previousView = this.view

        if (previousView === view) {
            this.view = null
        } else {
            throw IllegalStateException("Unexpected view! previousView = $previousView, view to unbind = $view")
        }
    }
}